File: javascript-node/src/runner.js
const resources = require("./resources");

module.exports = function(run) {
    for (let year = 2015; year <= 2020; year++) {
        console.log(`\n\n${year}\n`);
        for (let day = 1; day <= 25; day++) {
            let result = runDay(run, year, day);
            let padded = day.toString().padStart(2, " ");
            console.log(`${padded}   ${result}`);
        }
    }
}

function runDay(run, year, day) {
    try {
        const results = run(year, day);
        if (results == null) {
            return "Not implemented";
        } else {
            const answer1 = getAnswer(results[0], year, day, 1);
            const answer2 = getAnswer(results[1], year, day, 2);
            return `[${answer1}] [${answer2}]`;
        }
    } catch (e) {
        return "Error running";
    }
}

function getAnswer(result, year, day, part) {
    try {
        const answer = resources.readAnswer(year, day, part);
        if (answer == null) {
            return "?";
        } else {
            if (answer === result) {
                return "✓";
            } else {
                return "✖";
            }
        }
    } catch {
        return "⚠";
    }
}

File: kotlin/src/commonMain/kotlin/Runner.kt
fun runAll() {
    for (year in 2015..2020) {
        println("\n\n$year\n")
        (1..25)
            .map { runDay(year, it) }
            .forEach { (day, result) ->
                val padded = day.toString().padStart(2, ' ')
                println("$padded    $result")
            }
    }
}

fun runDay(year: Int, day: Int): Pair<Int, String> {
    return try {
        val results = run(year, day)
        if (results != null) {
            val answer1 = getAnswer(results.first, year, day, 1)
            val answer2 = getAnswer(results.second, year, day, 2)
            Pair(day, "[$answer1] [$answer2]")
        } else {
            Pair(day, "Not implemented")
        }
    } catch (e: Exception) {
        Pair(day, "Error running")
    }
}

fun getAnswer(result: String, year: Int, day: Int, part: Int): String {
    return try {
        val answer = Resources.readAnswerFile(year, day, part)
        if (answer != null) {
            if (result == answer) {
                "✓"
            } else {
                "✖"
            }
        } else {
            "?"
        }
    } catch (e: Exception) {
        "⚠"
    }
}